### PR TITLE
all: implement path based storage

### PR DESCRIPTION
**Pull request by sscodereth**
This PR implements the path based storage scheme. Note, it's a WIP PR, open it just want to ensure it's on the right track.

### CODE REVIEW

**Comment by holiman** [comment]
> for the deleted state object, we won't explicitly clean out the storage. So IIUC, nothing different will happen.

Hm, so in the current implementation, the storage trie will be left untouched after the deletion? If so, what will happen if

1. Contract A is created, via CREATE2
2. Contract A sets  slot `1` to `1`
4. Contract A is self-destructed
5. Contract A is resurrected. 

How do you avoid the accidental resurrection of the storage slots (which should now be cleared) if you haven't deleted them by-path?

**Comment by sscodereth** [comment]
> untouched

Actually it's OK to leave the deleted state data in the disk. Like in the case you described above, Contract A is recreated with an empty storage root hash, in this case the leaked storage data won't be accessed since **node hash is not matched**.

More specifically,  

In trie.Database.diffLayer, nodes are retrieved with internal key which is consisted with node hash. https://github.com/sscodereth/go-ethereum/blob/path-based-storage/trie/difflayer.go#L116

In trie.Database.diskLayer, nodes are retrieved with storage key(path), but data hash will still be compared with the requested one https://github.com/sscodereth/go-ethereum/blob/path-based-storage/trie/disklayer.go#L81

Let's go back to the scenario, the Contract A won't load any data for resolving an non-existent root node, so the left storage data won't be accessed at all.

File: eth/state_accessor.go [hunk 1]
```diff
@@ -125,9 +125,9 @@ func (eth *Ethereum) stateAtBlock(block *types.Block, reexec uint64, base *state
 		if err != nil {
 			return nil, fmt.Errorf("state reset after block %d failed: %v", current.NumberU64(), err)
 		}
-		database.TrieDB().Reference(root, common.Hash{})
+		//database.TrieDB().Reference(common.Hash{}, root, common.Hash{}, nil)
```
[the remark this answers is not in the document]
Comment (sscodereth) [reply eth/state_accessor.go:128]: This part(tracer) is still broken. Will fix it later.

**Comment by sscodereth** [comment]
> So -- we could choose to just store it as blobs, instead of storing as references to *cachedNode. That should improve gc a bit, and probably also is a bit safer (?)

The `cachedNode` should be super useful if the cached nodes can be reused by trie. Since RLP encoding and decoding is saved.

File: trie/disklayer.go [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+	}
+	path, hash := DecodeInternalKey(key)
+	blob, nodeHash := rawdb.ReadTrieNode(dl.diskdb, path)
+	if len(blob) == 0 || nodeHash != hash {
+		blob = rawdb.ReadArchiveTrieNode(dl.diskdb, hash)
+		if len(blob) != 0 {
+			triedbFallbackHitMeter.Mark(1)
+			triedbFallbackReadMeter.Mark(int64(len(blob)))
+		}
+	}
```
[the remark this answers is not in the document]
Comment (sscodereth) [reply trie/disklayer.go:146]: Yes, it's true. For all the non-existing entries, they will trigger an additional `ReadArchiveTrieNode`. It acts the fallback, mainly for the nodes upgraded from the legacy version.

**Comment by holiman** [comment]
> The cachedNode should be super useful if the cached nodes can be reused by trie. Since RLP encoding and decoding is saved.

But it's not used though, is it?

**Comment by holiman** [comment]
Happened during shutdown on one of the benchmarkers: 
```
Nov 23 22:11:06 bench03.ethdevops.io geth INFO [11-23|21:11:06.730] Got interrupt, shutting down...
Nov 23 22:11:06 bench03.ethdevops.io geth INFO [11-23|21:11:06.739] HTTP server stopped endpoint=[::]:8545
Nov 23 22:11:06 bench03.ethdevops.io geth INFO [11-23|21:11:06.740] IPC endpoint closed url=/datadir/geth/geth.ipc
Nov 23 22:11:06 bench03.ethdevops.io geth INFO [11-23|21:11:06.741] Stats daemon stopped
Nov 23 22:11:06 bench03.ethdevops.io geth WARN [11-23|21:11:06.741] Failed to retrieve stats server message err="read tcp 172.31.2.178:38228->52.9.113.136:443: use of closed network connection"
Nov 23 22:11:06 bench03.ethdevops.io geth INFO [11-23|21:11:06.744] Ethereum protocol stopped
Nov 23 22:11:06 bench03.ethdevops.io geth INFO [11-23|21:11:06.744] Transaction pool stopped 
Nov 23 22:11:06 bench03.ethdevops.io geth panic: reflect: call of reflect.Value.Type on zero Value
Nov 23 22:11:06 bench03.ethdevops.io geth goroutine 2821454975 [running]:
Nov 23 22:11:06 bench03.ethdevops.io geth reflect.Value.Type({0x0, 0x0, 0x1eaff60})
Nov 23 22:11:07 bench03.ethdevops.io geth reflect/value.go:2249 +0x12e
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/rlp.(*encbuf).encode(0x1eaff60, {0x0, 0x0})
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/rlp/encode.go:152 +0xb8
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/rlp.EncodeToBytes({0x0, 0x0})
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/rlp/encode.go:74 +0xe9
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/trie.(*cachedNode).rlp(0xc14b54fef8)
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/trie/database.go:197 +0x52
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/trie.(*diffLayer).Journal(0xc110d39ce0, 0x1c74660)
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/trie/journal.go:182 +0x2df
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/trie.(*diffLayer).Journal(0xc0c4da0150, 0x46db4a)
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/trie/journal.go:166 +0x67
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/trie.(*diffLayer).Journal(0xc0dfb65dc0, 0xd0)
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/trie/journal.go:166 +0x67
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/trie.(*diffLayer).Journal(0xc0ce9ba930, 0x3e)
... repeated a lot ...
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/trie/journal.go:166 +0x67
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/trie.(*diffLayer).Journal(0xc100e362a0, 0x46c857)
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/trie/journal.go:166 +0x67
Nov 23 22:11:07 bench03.ethdevops.io geth created by github.com/ethereum/go-ethereum/cmd/utils.StartNode.func1
Nov 23 22:11:07 bench03.ethdevops.io geth github.com/ethereum/go-ethereum/cmd/utils/cmd.go:90 +0x31f
```

**Comment by holiman** [comment]
Closing in favour of https://github.com/ethereum/go-ethereum/pull/24597

**Pull request closed without merging**